README.md
# ece243project
This is an assembly language project for the University of Toronto Second Year course ECE243:Computer Organization.

changelog.txt
Git commit history:
commit b8e6d39bb35f97385bd7b56ed061757a72eeef58
Author: A Wan <alewan@users.noreply.github.com>
Date:   Thu Apr 13 12:21:37 2017 -0400

    Made some fixes to hex display output, then paircoded with Alex to add beginning message.

commit 44176a3514e2d1b36fec5f075df9c28c3c2769eb
Author: A Wan <alewan@users.noreply.github.com>
Date:   Thu Apr 13 01:23:01 2017 -0400

    Added scorekeeping (output to 7seg HEX display),  and master game start (KEY0)

commit 3e31f044154b51bc3990a2ab94581da0149b94db
Author: MertTens <amertens114@gmail.com>
Date:   Wed Apr 12 17:20:16 2017 -0400

    FSR implementation
    
    The game works with the FSR now

commit 20a4c8d47741fcd0f52f13e5c7d66b986b9f6deb
Author: MertTens <amertens114@gmail.com>
Date:   Tue Apr 11 16:15:56 2017 -0400

    GPOI TEST
    
    the code to read whether or not the FSR has been pressed

commit 1b3ffa621ab617edaee6db1c7482d9dc4525baac
Author: A Wan <alewan@users.noreply.github.com>
Date:   Mon Apr 10 14:21:08 2017 -0400

    Changed VGA bar redraw method

commit dfc736cce5d73b7faa1198dc5638497a6ee92f61
Author: A Wan <alewan@users.noreply.github.com>
Date:   Mon Apr 10 13:38:23 2017 -0400

    Fixed a minor issue with right bar redraw

commit e44c5b4270f50af15eeab74f4df5960749dcfee4
Author: MertTens <amertens114@gmail.com>
Date:   Sun Apr 9 21:25:38 2017 -0400

    FSR
    
    added the arduino code for the FSR that will control the movement of
    one of the paddles

commit a812d238bb600685e2267b170799ad450e177e54
Author: MertTens <amertens114@gmail.com>
Date:   Sun Apr 9 21:21:05 2017 -0400

    Amplitude detection
    
    Added the code that is able to detect the maximum amplitude of some
    wave over time

commit d35234bcc4325a6110e2f8a080619e04601d9981
Author: A Wan <alewan@users.noreply.github.com>
Date:   Sun Apr 9 21:47:04 2017 -0400

    Overhauled bar drawing methodology, redrawing

commit 2303f51d0059fce42c01e63e9278c4fd2e5e54f5
Author: A Wan <alewan@users.noreply.github.com>
Date:   Sun Apr 9 21:02:19 2017 -0400

    Adjusted bar drawing and tracking

commit 90e231b89fb687dd811677f049a271aeb3c228e7
Author: alewan <alewan@users.noreply.github.com>
Date:   Sun Apr 9 20:31:19 2017 -0400

    Fixed minor issues with screen border detection

commit 6abe9844c60d92b6af9e4470df1eb16014cd1208
Author: A Wan <alewan@users.noreply.github.com>
Date:   Sun Apr 9 20:15:44 2017 -0400

    Added VGA border detection

commit 74413996790a1c236073726c7aecd0450869a572
Merge: 6e32f3d 9b15813
Author: A Wan <alewan@users.noreply.github.com>
Date:   Sat Apr 8 11:26:12 2017 -0400

    Resolved a merge conflict

commit 6e32f3d0c51b1ee77328d2a6c078bae01c9ab4ea
Author: A Wan <alewan@users.noreply.github.com>
Date:   Sat Apr 8 11:22:31 2017 -0400

    Minor changes and reverted key reset because it caused practical issues

commit 9b15813c603286ecc4dd03bf571cfa188fdcdd2b
Author: A Wan <alewan@users.noreply.github.com>
Date:   Thu Apr 6 13:32:24 2017 -0400

    Added a master reset   (interrupt from KEY[0])

commit 49c8ae0e396876f7c08ace00b7c52e78019d6cb8
Author: A Wan <aleksei.wan@gmail.com>
Date:   Thu Apr 6 11:46:03 2017 -0400

    Initial commit


Before the project was in Git:
April 5, 2017
Aleksei: Wrote modular code to draw additional game elements on the VGA and keep track of them. The functions were also written in a way to support code reuse for I/O integration. Also added code for timer interrupt driven motion/re-draw.
Alex: Wrote basic code for testing audio. Elaborated on this code by adding amplitude detection. A document containing the used registers and their exact purpose was included.

March 30, 2017
Aleksei: Set up structure and formatting of the files, created basic interactions with VGA including resetting the screen and other core functionality.


GPOI_test.s
.equ ADDR_JP2PORT, 0xFF200070
.equ ADDR_JP2PORT_DIR, 0xFF200074



.section .text
.global _start:
_start:
  movia r2,ADDR_JP2PORT_DIR
  stwio r0,0(r2)  # Set direction of all 32-bits to input 

main:
  movia r2,ADDR_JP2PORT
  ldwio r3,(r2)   # Read value from pins
  andi r3, r3, 0x01

  br main

nios_audio.s
.equ AUDIO, 0xFF203040 	   # (From DESL website > NIOS II > devices)
.equ JTAG_UART, 0XFF201000

.global _start
_start:
	#max ampliltude initialize
	movi r9, 0
	#number of samples needed to chech before update max amp
	movi r17, 384
	#local max amplitude withing the 384 samples
	movi r18, 0
	#the loop counter
	movi r15, 0
	
	#determine when to print to jtag
	movi r11, 0
	movi r12, 50
	main:
	movia r2, AUDIO
	ldwio r3, 4(r2)
	andi r3, r3, 0xff
	beq r3, r0, main
	ldwio r3, 8(r2)
	stwio r3, 8(r2)
	ldwio r3, 12(r2)
	stwio r3, 12(r2)
	
	#update the local max amplitude
	bgt r3, r18, updateLocalAmp
	returnFromLocalAmp:
	#incriment the loop counter
	addi r15, r15, 1
	
	#if enough loops have happened, update max amp and reset local amp
	bge r15, r17, updateMaxAmp
	returnFromMaxAmp:
	

	srli r10, r9, 23
	br main
	
	
	
	updateLocalAmp:
	mov r18, r3
	br returnFromLocalAmp
	
	updateMaxAmp:
	mov r9, r18
	movi r18, 0
	movi r15, 0
	
	addi r11, r11, 1
	bge r11, r12, printJTAG
	
	br returnFromMaxAmp
	
	
	printJTAG:
	mov r11, r0
	movia r20, JTAG_UART
	
	movui r19, 0x1b
	stwio r19, 0(r20)
	movui r19, 0x5b
	stwio r19, 0(r20)
	movui r19, 0x32
	stwio r19, 0(r20)
	movui r19, 0x4a
	stwio r19, 0(r20)
	
	movia r19, 0x0600000
	blt r9, r19, printS
	
	movia r19, 0x50000000
	blt r9, r19, printM
	
	
	movia r20, JTAG_UART
	movui r19, 0x4c
	stwio r19, 0(r20)
	br returnFromMaxAmp
	
	printS:
	movia r20, JTAG_UART
	movui r19, 0x53
	stwio r19, 0(r20)
	br returnFromMaxAmp
	
	printM:
	movia r20, JTAG_UART
	movui r19, 0x4d
	stwio r19, 0(r20)
	br returnFromMaxAmp
	

project.s
#Conventions
#Functions will use caller saved registers r8, r9 first then callee saved registers
#Recall: VGA pixel offset requires adding 2*x + 1024*y
#"Global Variables"
#r22 is X location of center of the ball, r20 is X directionality
#r23 is Y location of center of the ball, r21 is Y directionality
#r18 is the new top of the left bar
#r19 is the new top of the right bar
#r16 (Player 1) and r17 (Player 2) are the scores

#External Devices
.equ JTAG_UART, 0xFF201000
.equ old_time, 0xF08002FA #500M in Hex
.equ redraw_time, 0x00200000
.equ stuffedcow_time, x00100000
.equ VGA_PIXEL_BASE, 0x08000000
.equ VGA_CHAR_BASE, 0x09000000
.equ HEX_DISPLAY, 0xFF200020 #every 8 bits is a new HEX display until bit 30 (6-0 H0, 14-8 H1, 22-16 H2, 30-24 H3)
.equ AUDIO, 0XFF203040
.equ IRQ_SETUP, 0x00000003 #line0 for timer, line1 for keys
.equ TIMER1, 0xFF202000
.equ ADDR_JP2PORT, 0xFF200070
.equ ADDR_JP2PORT_DIR, 0xFF200074
.equ KEYS, 0xFF200050

#VGA Definitions
.equ XMAX, 320
.equ YMAX, 240
.equ BALL_START_X, 160
.equ BALL_START_Y, 120
.equ BAR1_START_X, 1
.equ BAR2_START_X, 316
.equ BAR1_START_Y, 114
.equ BAR2_START_Y, 114
.equ BAR_SIZE, 35
.equ WHITE, 0xFFFF
.equ GREY, 0x8410
.equ BLACK, 0x0000
.equ RED, 0xF800
.equ GREEN, 0x07E0
.equ BLUE, 0x001F
.equ PURPLE, 0xF81F

#Hex Definitions
.equ HEX_DISPLAY_VAL0, 0x3F
.equ HEX_DISPLAY_VAL1, 0x06
.equ HEX_DISPLAY_VAL2, 0x5B
.equ HEX_DISPLAY_VAL3, 0x4F
.equ HEX_DISPLAY_VAL4, 0x66
.equ HEX_DISPLAY_VAL5, 0x6D
.equ HEX_DISPLAY_VAL6, 0x7D
.equ HEX_DISPLAY_VAL7, 0x07
.equ HEX_DISPLAY_VAL8, 0x7F
.equ HEX_DISPLAY_VAL9, 0x6F
.equ HEX_DISPLAY_VALA, 0b1110111

#bytes
#beginning_message: .byte 0x50, 0x72, 0x65, 0x73, 0x73, 0x20, 0x4b, 0x45, 0x59, 0x30, 0x0d, 0x0a


#TO SAVE IN THE ISR
#r8
#r9
#r10
#r11
#r12
#r13

#ISR
.section .exceptions,"ax"
ISR:
	rdctl et, ipending #check ipending
	#look for timer interrupt
	andi et, et, 0x1
	bne et, r0, TIMER_ISR
	rdctl et, ipending
	#look for key interrupt
	andi et, et, 0x2
	beq et, r0, END_ISR

KEYS_ISR:
	#Start of keys interrupt handling
	#acknowledge interrupt
	movia et, KEYS
	stwio r0, 12(et)
	#prevent future keys interrupts
	stwio r0, 8(et)
	#Exit to break the loop
	eret
	#End of keys interrupt handling

TIMER_ISR:
	#Start of Timer Interrupt Handling
	addi sp, sp, -40
	stw r8,(sp)
	stw r9,4(sp)
	stw r10,8(sp)
	stw r11,12(sp)
	stw r12,16(sp)
	stw r13,20(sp)
	stw r14,24(sp)
	stw r4,28(sp)
	stw r5,32(sp)
	stw r6,36(sp)
	

	#check the fsr
	movia r11, ADDR_JP2PORT_DIR
	stwio r0, 0(r11)
	movia r11, ADDR_JP2PORT
	ldwio r14, 0(r11)
	movia r12, 0xffffffff
	bne r14, r12, addr19
	br subr19
	retFromFSR:
	call REDRAW_ICON
	call REDRAW_BAR
	call UPDATE_SCORE
	#acknowledge interrupt and reset timer
	movia et, TIMER1
	stwio r0, (et)
	movui r8, 0b101 #Enable start, no CTS, with interrupt
	stwio r8, 4(et)
	ldw r8,(sp)
	ldw r9,4(sp)
	ldw r10,8(sp)
	ldw r11,12(sp)
	ldw r12,16(sp)
	ldw r13,20(sp)
	ldw r14,24(sp)
	ldw r4,28(sp)
	ldw r5,32(sp)
	ldw r6,36(sp)
	addi sp, sp, 40
	br END_ISR
	#End of Timer Interrupt Handling

END_ISR:
	addi ea, ea, -4
	eret

.section .text
.global _start
#Main code
_start:
#BEGIN SETUP
movia sp, 0x04000000

movui r4, BLACK
call RESET_VGA

#Initialize scores to zero
mov r16, r0
mov r17, r0

#Draw first bar
movui r4, BLUE
addi r5, r0, BAR1_START_X
addi r6, r0, BAR1_START_Y
call DRAW_BAR
addi r18,r0,BAR1_START_Y

#Draw second bar
movui r4, BLUE
addi r5, r0, BAR2_START_X
addi r6, r0, BAR2_START_Y
call DRAW_BAR
addi r19,r0,BAR2_START_Y

#Draw the ball
movui r4, GREEN
addi r5, r0, BALL_START_X
addi r6, r0, BALL_START_Y
call DRAW_ICON
movui r22, BALL_START_X
movui r23, BALL_START_Y
addi r20,r0,1
addi r21,r0,1

call UPDATE_SCORE
#END SETUP

#ISR_SETUP
	#Catch any previous interrupts and clear them
	movia r8,TIMER1
	stwio r0,(r8)
	stwio r0,4(r8)
	movia r8,KEYS
	stwio r0,8(r8)
	stwio r0,12(r8)
	
	#Enable interrups
	movi r9, IRQ_SETUP
	wrctl ienable, r9 #ienable = ctl3
	movi r9, 0b1
	wrctl status, r9 #Enable PIE

	#Setup keys and then wait for keypress to start the game
	movia r8, KEYS
	movia r9,0x1
	stwio r9,8(r8)
	stwio r0,12(r8)
	#MESSGAE TO VGA BEFORE START OF GAME
	#bytes
#beginning_message: .byte 0x50, 0x72, 0x65, 0x73, 0x73, 0x20, 0x4b, 0x45, 0x59, 0x30, 0x0d, 0x0a
	movia r8, VGA_CHAR_BASE
	movi r7, 0x50
	stbio r7, 291(r8)
	movi r7, 0x72
	stbio r7, 292(r8)
	movi r7, 0x65
	stbio r7, 293(r8)
	movi r7, 0x73
	stbio r7, 294(r8)
	movi r7, 0x73
	stbio r7, 295(r8)
	movi r7, 0x20
	stbio r7, 296(r8)
	movi r7, 0x4b
	stbio r7, 297(r8)
	movi r7, 0x45
	stbio r7, 298(r8)
	movi r7, 0x59
	stbio r7, 299(r8)
	movi r7, 0x30
	stbio r7, 300(r8)

	
	
	WAIT_FOR_KEYPRESS: br WAIT_FOR_KEYPRESS
		movia r8, VGA_CHAR_BASE
	movi r7, 0x20
	stbio r7, 291(r8)
	stbio r7, 292(r8)
	stbio r7, 293(r8)
	stbio r7, 294(r8)
	stbio r7, 295(r8)
	stbio r7, 296(r8)
	stbio r7, 297(r8)
	stbio r7, 298(r8)
	stbio r7, 299(r8)
	stbio r7, 300(r8)


	#Initialize counter value
	movia r8, TIMER1
	movui r9, %lo(redraw_time)
	stwio r9, 8(r8)
	movui r9, %hi(redraw_time)
	stwio r9, 12(r8)

	stwio r0, (r8) #reset TIMER1
	movui r9, 0b101 #Enable start, no CTS, enable interrupt
	stwio r9, 4(r8)
#END ISR_SETUP

#number of samples needed
movi r8, 384

#local max within the loop
movi r9, 0

#the loop counter
movi r10, 0


#TO SAVE IN THE ISR
#r8
#r9
#r10
#r11
#r12
#r13
#r14

gameLoop:

checkAudio:
#check if the fifo has anything in it
movia r11, AUDIO
ldwio r12, 4(r11)
andi r12, r12, 0xFF
beq r12, r0, checkAudio

#echo audio to the speakers (stored in r12)
ldwio r12, 8(r11)
stwio r12, 8(r11)
ldwio r12, 12(r11)
stwio r12, 12(r11)

#update the local max amplitude
bgt r12, r9, updateLocalAmp
returnFromLocalAmp:

#incriment the loop counter
addi r10, r10, 1

#if enough loops, then update the global max amplitude
bge r10, r8, updateMaxAmp
returnFromMaxAmp:


br gameLoop

#add to r19
addr19:
movi r12, YMAX
subi r12, r12, BAR_SIZE
bgt r19, r12, retFromFSR
addi r19, r19, 5
br retFromFSR

#sub from r19
subr19:
blt r19, r0, retFromFSR
subi r19, r19, 5
br retFromFSR

#update the max amplitude within the loop
updateLocalAmp:
mov r9, r12
br returnFromLocalAmp 

updateMaxAmp:
movi r10, 0
#temp register for shifting
mov r13, r9
srli r13, r13, 23
movi r9, 0

#maybe later here make it stay below 240
mov r18, r13
br returnFromMaxAmp

#END: br END not necessary since the game loops in game loop

#Function to reset VGA to colour specified in r4
RESET_VGA:
	#Prologue
	addi sp, sp, -8
	stw r10,0(sp)
	stw r11,4(sp)

	#Initialization
	movia r8, VGA_PIXEL_BASE
	mov r11, r0 #ycount
	
	#Core
	VGAR_Y_LOOP:
		mov r10, r0 #xcount
		movui r9, XMAX
		VGAR_X_LOOP:
			sthio r4,(r8)
			addi r8,r8,2
			addi r10, r10, 1
			blt r10, r9, VGAR_X_LOOP
		subi r8, r8, 640 #2*XMAX
		addi r8, r8, 1024
		addi r11, r11, 1
		movui r9, YMAX
		blt r11, r9, VGAR_Y_LOOP
	
	#Epilogue
	ldw r10,0(sp)
	ldw r11,4(sp)
	addi sp, sp, 8
	ret

#Draw over the column with color specified in r4, leftmost x location in r5
CLEAR_BAR_COLUMN:
	#Prologue
	addi sp, sp, -4
	stw ra,0(sp)
	
	#Initialization
	#Adjust arguments to center
	muli r5, r5, 2
	add r6,r5,r0
	movia r5,VGA_PIXEL_BASE
	add r5,r5,r6
	
	#Core
	movui r6, 240
CLEAR_BAR_DRAWING_LOOP:
	call DRAW_BAR_SECTION
	addi r5,r5,1024
	addi r6,r6,-1
	bgt r6,r0,CLEAR_BAR_DRAWING_LOOP

	#Epliogue
	ldw ra,(sp)
	addi sp, sp, 4
	ret

#Draw an arbitrarily sized bar with color specified in r4, leftmost x location in r5, top y location in r6
DRAW_BAR:
	#Prologue
	addi sp, sp, -4
	stw ra,0(sp)
	
	#Initialization
	#Adjust arguments to center
	muli r5, r5, 2
	muli r6, r6, 1024
	add r6,r5,r6
	movia r5,VGA_PIXEL_BASE
	add r5,r5,r6
	
	#Core
	movui r6, BAR_SIZE
BAR_DRAWING_LOOP:
	call DRAW_BAR_SECTION
	addi r5,r5,1024
	addi r6,r6,-1
	bgt r6,r0,BAR_DRAWING_LOOP

	#Epliogue
	ldw ra,(sp)
	addi sp, sp, 4
	ret

#Function to draw a 3x1 line at a location specified by color=r4,location=r5
DRAW_BAR_SECTION:
	#Prologue (N/A)
	
	#Draw 3 bits
	sthio r4, (r5)
	sthio r4, 2(r5)
	sthio r4, 4(r5)
	
	#Epilogue (N/A)
	ret

#Function to redraw the bar
REDRAW_BAR:
	#Prologue
	addi sp, sp, -4
	stw ra,(sp)

	#Initialization
	movia r8, VGA_PIXEL_BASE

	#Redraw over old location
	movui r4, BLACK
	movui r5, BAR1_START_X
	call CLEAR_BAR_COLUMN
	movui r5, BAR2_START_X
	call CLEAR_BAR_COLUMN

	#Draw in new location
	movui r4, BLUE
	movui r5, BAR1_START_X
	mov r6, r18
	call DRAW_BAR
	movui r4, BLUE
	movui r5, BAR2_START_X
	mov r6, r19
	call DRAW_BAR

	#Epliogue
	ldw ra,(sp)
	addi sp, sp, 4
	ret


#Function to draw 3x3 icon at a location specified by x=r5, y=r6, color=r4
DRAW_ICON:
	#Prologue (N/A)
	
	#Initialization
	movia r8, VGA_PIXEL_BASE
	
	#Core
	#Adjust r8 to center
	muli r5, r5, 2
	muli r6, r6,1024
	add r8, r8, r5
	add r8, r8, r6
	addi r8, r8, -1026 #Start at top LH corner (center-1026)
	
	#Top three bits
	sthio r4, (r8)
	sthio r4, 2(r8)
	sthio r4, 4(r8)
	
	#Middle three bits
	sthio r4, 1024(r8)
	sthio r4, 1026(r8)
	sthio r4, 1028(r8)	
	
	#Bottom three bits
	sthio r4, 2048(r8)
	sthio r4, 2050(r8)
	sthio r4, 2052(r8)
	
	#Epliogue (N/A)
	ret


#Function to re-draw 3x3 icon at a location
REDRAW_ICON:
	#Prologue
	addi sp, sp, -4
	stw ra,(sp)

	#Initialization
	movia r8, VGA_PIXEL_BASE

	#Redraw over old location
	movui r4, BLACK
	mov r5, r22
	mov r6, r23
	call DRAW_ICON
	
	#Correct r22, r23 with border detection
	#Check X direction
	bgt r20,r0,MOVING_RIGHT
MOVING_LEFT:
	addi r22, r22, -2 
	br Y_DIR_CHECK
MOVING_RIGHT: 
	addi r22, r22, 2
Y_DIR_CHECK:
	bgt r21, r0, MOVING_DOWN
MOVING_UP:
	addi r23,r23, -2
	br BORDER_CHECK
MOVING_DOWN:
	addi r23, r23, 2
BORDER_CHECK:
	#Check for the left bar
	movui r4, BAR1_START_X
	addi r4,r4,4
	bgt r22,r4,DONE_CHECK_HIT_B1
	blt r23,r18, DONE_CHECK_HIT_B1
	addi r4,r18,BAR_SIZE
	bgt r23,r4,DONE_CHECK_HIT_B1
	#Getting here implies a hit on the bar
	xori r20,r20,1
	br CHECK_Y_DIR
DONE_CHECK_HIT_B1:
	#Check for the right bar
	movui r4, BAR2_START_X
	subi r4,r4,1
	blt r22,r4,DONE_CHECK_HIT_B2
	blt r23,r19, DONE_CHECK_HIT_B2
	addi r4,r19,BAR_SIZE
	bgt r23,r4,DONE_CHECK_HIT_B2
	#Getting here implies a hit on the bar
	xori r20,r20,1
	br CHECK_Y_DIR
DONE_CHECK_HIT_B2:
	#Check for the wall
	movui r4,XMAX
	#Check X border, Y border
	bge r22,r4,CHANGE_X_DIR_GOALP1
	ble r22,r0,CHANGE_X_DIR_GOALP2
	br CHECK_Y_DIR
CHANGE_X_DIR_GOALP1:
	addi r16,r16,1
	br CHANGE_X_DIR
CHANGE_X_DIR_GOALP2:
	addi r17,r17,1
CHANGE_X_DIR:
	xori r20,r20,1
CHECK_Y_DIR:
	movui r4,YMAX
	bge r23,r4,CHANGE_Y_DIR
	ble r23,r0,CHANGE_Y_DIR
	br END_CHANGE_DIR
CHANGE_Y_DIR:
	xori r21,r21,1

END_CHANGE_DIR:
	#Core
	movui r4, GREEN
	mov r5, r22
	mov r6, r23
	call DRAW_ICON
	
	#Epliogue
	ldw ra,(sp)
	addi sp, sp, 4
	ret

#Function to write a value in r16 to the 7-seg display HEX0 and r17 to HEX1
UPDATE_SCORE:
	#Prologue (N/A)

	#Initialization
	movia r8, HEX_DISPLAY
	
	#Core
	#Check value in r16
	beq r16, r0, DRAW0_P1
	addi r9,r0,1
	beq r16, r9, DRAW1_P1
	addi r9,r9,1
	beq r16, r9, DRAW2_P1
	addi r9,r9,1
	beq r16, r9, DRAW3_P1
	addi r9,r9,1
	beq r16, r9, DRAW4_P1
	addi r9,r9,1
	beq r16, r9, DRAW5_P1
	addi r9,r9,1
	beq r16, r9, DRAW6_P1
	addi r9,r9,1
	beq r16, r9, DRAW7_P1
	addi r9,r9,1
	beq r16, r9, DRAW8_P1
	addi r9,r9,1
	beq r16, r9, DRAW9_P1
	addi r9,r9,1
	beq r16, r9, DRAWA_P1

#Update the score
DRAW0_P1:
	movui r9, HEX_DISPLAY_VAL0
	br UPDATE_P2
DRAW1_P1:
	movui r9, HEX_DISPLAY_VAL1
	br UPDATE_P2
DRAW2_P1:
	movui r9, HEX_DISPLAY_VAL2
	br UPDATE_P2
DRAW3_P1:
	movui r9, HEX_DISPLAY_VAL3
	br UPDATE_P2
DRAW4_P1:
	movui r9, HEX_DISPLAY_VAL4
	br UPDATE_P2
DRAW5_P1:
	movui r9, HEX_DISPLAY_VAL5
	br UPDATE_P2
DRAW6_P1:
	movui r9, HEX_DISPLAY_VAL6
	br UPDATE_P2
DRAW7_P1:
	movui r9, HEX_DISPLAY_VAL7
	br UPDATE_P2
DRAW8_P1:
	movui r9, HEX_DISPLAY_VAL8
	br UPDATE_P2
DRAW9_P1:
	movui r9, HEX_DISPLAY_VAL9
	br UPDATE_P2
DRAWA_P1:
	movui r9, HEX_DISPLAY_VALA
	stwio r9,(r8)
	#GAMEOVER
	movia ea, _start
	eret

UPDATE_P2:
	stwio r9,(r8) #Sending P1 score to hex display
#Check value in r17
	beq r17, r0, DRAW0_P2
	addi r9,r0,1
	beq r17, r9, DRAW1_P2
	addi r9,r9,1
	beq r17, r9, DRAW2_P2
	addi r9,r9,1
	beq r17, r9, DRAW3_P2
	addi r9,r9,1
	beq r17, r9, DRAW4_P2
	addi r9,r9,1
	beq r17, r9, DRAW5_P2
	addi r9,r9,1
	beq r17, r9, DRAW6_P2
	addi r9,r9,1
	beq r17, r9, DRAW7_P2
	addi r9,r9,1
	beq r17, r9, DRAW8_P2
	addi r9,r9,1
	beq r17, r9, DRAW9_P2
	addi r9,r9,1
	beq r17, r9, DRAWA_P2
	
DRAW0_P2:
	movui r9, HEX_DISPLAY_VAL0
	br END_UPDATE_SCORE
DRAW1_P2:
	movui r9, HEX_DISPLAY_VAL1
	br END_UPDATE_SCORE
DRAW2_P2:
	movui r9, HEX_DISPLAY_VAL2
	br END_UPDATE_SCORE
DRAW3_P2:
	movui r9, HEX_DISPLAY_VAL3
	br END_UPDATE_SCORE
DRAW4_P2:
	movui r9, HEX_DISPLAY_VAL4
	br END_UPDATE_SCORE
DRAW5_P2:
	movui r9, HEX_DISPLAY_VAL5
	br END_UPDATE_SCORE
DRAW6_P2:
	movui r9, HEX_DISPLAY_VAL6
	br END_UPDATE_SCORE
DRAW7_P2:
	movui r9, HEX_DISPLAY_VAL7
	br END_UPDATE_SCORE
DRAW8_P2:
	movui r9, HEX_DISPLAY_VAL8
	br END_UPDATE_SCORE
DRAW9_P2:
	movui r9, HEX_DISPLAY_VAL9
	br END_UPDATE_SCORE
DRAWA_P2:
	movui r9, HEX_DISPLAY_VALA
	slli r9,r9,8
	ldwio r8,(r8)
	or r9,r8,r9
	movia r8,HEX_DISPLAY
	stwio r9,(r8)
	#GAMEOVER
	movia ea, _start
	eret

END_UPDATE_SCORE:
	slli r9,r9,8
	ldwio r8,(r8)
	or r9,r8,r9
	movia r8,HEX_DISPLAY
	stwio r9,(r8) #Sending P2 score to hex display
	#Epilogue (N/A)
	ret
